dts: arm: renesas: ra: Separate the OFS memory into individual regions

jeremydick opened the pull request:
Create separate memory regions for each OFS register.
With a single region the linker will gap fill the load segment with zeros between each option setting section that gets placed in the region when generating the .elf file.

This fixes a regression introduced in https://github.com/zephyrproject-rtos/zephyr/pull/93137, which can result in writing zeros to any areas of the OFS that aren't explicitly set when flashing the elf file, potentially permanently lock the flash.


An example seen on the RA6M4

```
> readelf -l build/zephyr/zephyr.elf

Elf file type is EXEC (Executable file)
Entry point 0x70e1
There are 7 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x013248 0x00013130 0x00013130 0x00008 0x00008 R   0x4
  LOAD           0x000118 0x00000000 0x00000000 0x1b7e4 0x1b7e4 RWE 0x8
  LOAD           0x01b900 0x20000000 0x0001b7e4 0x01208 0x01208 RWE 0x8
  LOAD           0x01cb08 0x0001c9ec 0x0001c9ec 0x00004 0x00004 RW  0x4
  LOAD           0x01cb0c 0x0100a100 0x0100a100 0x001cc 0x001cc R   0x4
  LOAD           0x000000 0x20001208 0x20001208 0x00000 0x04510 RW  0x8
  TLS            0x0143e8 0x000142d0 0x000142d0 0x00000 0x0002c R   0x4

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx
   01     rom_start text .ARM.exidx initlevel device_area sw_isr_table _static_thread_data_area adc_driver_api_area flash_driver_api_area gpio_driver_api_area i2c_driver_api_area led_driver_api_area pwm_driver_api_area rtc_driver_api_area sensor_driver_api_area spi_driver_api_area bbram_driver_api_area clock_control_driver_api_area regulator_driver_api_area uart_driver_api_area input_callback_area shell_area shell_root_cmds_area shell_subcmds_area shell_dynamic_subcmds_area cfb_font_area rodata
   02     .ramfunc datas device_states pm_device_slots_area k_heap_area k_msgq_area k_event_area
   03     .last_section
   04     .option_setting_ofs0 .option_setting_dualsel .option_setting_ofs1_sec .option_setting_banksel_sec .option_setting_bps_sec .option_setting_pbps_sec .option_setting_ofs1_sel .option_setting_banksel_sel .option_setting_bps_sel
   05     bss noinit
   06     tbss
```

load segment 4 will write 0x1cc byes to the OFS at 0x0100a100, the contents of which are:

```
> hexdump -s 0x01cb0c -n 0x1cc -C build/zephyr/zephyr.elf
0001cb0c  ff ff ff ff 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0001cb2c  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0001cc0c  ff fd ff ff 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0001cc1c  ff ff ff ff 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0001cc2c  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0001cc4c  ff ff ff ff ff ff ff ff  ff ff ff ff 00 00 00 00  |................|
0001cc5c  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0001cc6c  ff ff ff ff ff ff ff ff  ff ff ff ff 00 00 00 00  |................|
0001cc7c  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0001cc8c  f8 f8 ff ff 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0001cc9c  ff ff ff ff 00 00 00 00  00 00 00 00 00 00 00 00  |................|
0001ccac  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0001cccc  ff ff ff ff ff ff ff ff  ff ff ff ff              |............|
0001ccd8
```

thenguyenyf reviewed (dismissed):
Pending until complete discussion. @khoa-nguyen-18 . Please take a look.

khoa-nguyen-18 commented:
Hi, apologies for my late response. I have a concern about using this approach. On some MCUs, there are numerous OFS regions. How should we handle the extensive build logs generated in such cases?

<img width="712" height="282" alt="image" src="https://github.com/user-attachments/assets/92b5d6ed-906e-427a-a28b-f7c64ec3c438" />


FYI> I have verified using J-Flash that both the RA6M4 ELF and HEX files allocate nothing to the OFS areas that aren't explicitly set.

<img width="702" height="555" alt="image" src="https://github.com/user-attachments/assets/996c9c47-de77-4f5d-b4da-bd536dc438e6" />

khoa-nguyen-18 commented on dts/arm/renesas/ra/ra2/ra2l1.dtsi:
```diff
@@ -542,15 +542,15 @@
 
 		option_setting_ofs_conf: option-setting-ofs-conf@1010010 {
-			compatible = "zephyr,memory-region";
+			compatible = "renesas,ofs-memory";
```
Should we remove the compatible property for these parent OFS nodes and just use `renesas,ofs-memory` for any node that can't be used by `zephyr,memory-region`, such as the child nodes of `option_setting_ofs_in_rom`?

jeremydick replied:
I think that is a good suggestion

khoa-nguyen-18 commented on dts/arm/renesas/ra/ra2/ra2l1.dtsi:
```diff
@@ -542,15 +542,15 @@
 			option_setting_osis: option-setting-osis@1010018 {
-				compatible = "renesas,ofs-memory";
+				compatible = "zephyr,memory-region";
+				zephyr,memory-region = "OPTION_SETTING_OSIS_MEMORY";
```
Should we shorten the name of the OFS region, maybe to `OPTION_SETTING_OSIS` or `OFS_OSIS`?

jeremydick replied:
Yes, I'll shorten all the region names

khoa-nguyen-18 reviewed (commented):
I think this is still a good way to define the OFS region. The PR is good to go, but I have some information that needs to be confirmed. cc @duynguyenxa, @thenguyenyf

khoa-nguyen-18 reviewed (dismissed):
good work

khoa-nguyen-18 commented:
Hi @jeremydick, could you please rebase to resolve the conflicts?
FYI> We are planning to upstream a major update. It would be great if you could help us finish this PR soon so we can integrate it internally before upstreaming.

jeremydick committed:
dts: arm: renesas: ra: Separate the OFS memory into individual regions

Create separate memory regions for each OFS register. With a single
region the linker will gap fill the load segment with zeros between
each option setting section that gets placed in the region when
generating the .elf file.

Signed-off-by: Jeremy Dick <jdick@pivotint.com>

thenguyenyf commented on dts/arm/renesas/ra/ra2/ra2l1.dtsi:
```diff
@@ -549,15 +549,14 @@
 		};
 
 		option_setting_ofs_conf: option-setting-ofs-conf@1010010 {
```
The entire node `option_setting_ofs_conf: option-setting-ofs-conf@1010010` should be deleted (not only the nodelabel) and all its children should be moved to `/{soc{}}`.

thenguyenyf reviewed (approved):
Remain few nits, but we can keep it right now and improve in the future
